beacon-chain/core/epoch/epoch_processing.go

```diff
@@ -0,0 +1,107 @@
+func ProcessReceipt(state *pb.BeaconState) *pb.BeaconState {
+	newState := proto.Clone(state).(*pb.BeaconState)
+	for _, receiptRoot := range state.CandidatePowReceiptRoots {
+		if receiptRoot.VoteCount*2 > params.BeaconConfig().PowReceiptRootVotingPeriod {
+			newState.ProcessedPowReceiptRootHash32 = receiptRoot.CandidatePowReceiptRootHash32
+		}
+	}
+	newState.CandidatePowReceiptRoots = make([]*pb.CandidatePoWReceiptRootRecord, 0)
+	return newState
+}
+
+// ProcessJustification processes for justified slot by comparing
+// epoch boundary balance and total balance.
+//
+// Spec pseudocode definition:
+//    Set state.previous_justified_slot = state.justified_slot.
+//    Set state.justification_bitfield = (state.justification_bitfield * 2) % 2**64.
+//    Set state.justification_bitfield |= 2 and state.justified_slot =
+//    state.slot - 2 * EPOCH_LENGTH if 3 * previous_epoch_boundary_attesting_balance >= 2 * total_balance.
+//    Set state.justification_bitfield |= 1 and state.justified_slot =
+//    state.slot - 1 * EPOCH_LENGTH if 3 * this_epoch_boundary_attesting_balance >= 2 * total_balance.
+func ProcessJustification(
+	state *pb.BeaconState,
+	thisEpochBoundaryAttestingBalance uint64,
+	prevEpochBoundaryAttestingBalance uint64,
+	totalBalance uint64) *pb.BeaconState {
+
+	newState := proto.Clone(state).(*pb.BeaconState)
+	newState.PreviousJustifiedSlot = state.JustifiedSlot
+	// Shifts all the bits over one to create a new bit for the recent epoch.
+	newState.JustificationBitfield = state.JustificationBitfield * 2
+
+	// If prev prev epoch was justified then we ensure the 2nd bit in the bitfield is set,
+	// assign new justified slot to 2 * EPOCH_LENGTH before.
+	if 3*prevEpochBoundaryAttestingBalance >= 2*totalBalance {
+		newState.JustificationBitfield |= 2
+		newState.JustifiedSlot = state.Slot - 2*params.BeaconConfig().EpochLength
+	}
+
+	// If this epoch was justified then we ensure the 1st bit in the bitfield is set,
+	// assign new justified slot to 1 * EPOCH_LENGTH before.
+	if 3*thisEpochBoundaryAttestingBalance >= 2*totalBalance {
+		newState.JustificationBitfield |= 1
+		newState.JustifiedSlot = state.Slot - 1*params.BeaconConfig().EpochLength
+	}
+	return newState
+}
+
+// ProcessFinalization processes for finalized slot by checking
+// consecutive justified slots.
+//
+// Spec pseudocode definition:
+//   Set state.finalized_slot = state.previous_justified_slot if any of the following are true:
+//		state.previous_justified_slot == state.slot - 2 * EPOCH_LENGTH and state.justification_bitfield % 4 == 3
+//		state.previous_justified_slot == state.slot - 3 * EPOCH_LENGTH and state.justification_bitfield % 8 == 7
+//		state.previous_justified_slot == state.slot - 4 * EPOCH_LENGTH and state.justification_bitfield % 16 in (15, 14)
+func ProcessFinalization(state *pb.BeaconState) *pb.BeaconState {
+	epochLength := params.BeaconConfig().EpochLength
+
+	newState := proto.Clone(state).(*pb.BeaconState)
```

Comment:
Do we need to proto clone every single time though? In my PR, since all the block operations will be done sequentially in the same function `ProcessBlock`, I only clone once before calling all the operations and pass in the cloned state for all the transformations. Finally, I return the cloned state.

Reply:
Yeah, I'm just being extra careful, but it's probably unnecessary to clone every time. We can just clone at the beginning of  epoch processing go routine at the blockchain service layer

Reply:
Its not needed imo, also it will be pretty inefficient to clone the state for each small function. The cloning of state should be handled in processing routines not the justification/finalization/other functions

Reply:
Sounds good. Will remove them 👍